homeassistant/helpers/service.py

```diff
@@ -581,31 +581,36 @@ async def async_get_all_descriptions(
     descriptions_cache: dict[
         tuple[str, str], dict[str, Any] | None
     ] = hass.data.setdefault(SERVICE_DESCRIPTION_CACHE, {})
-    services = hass.services.async_services()
+
+    # We don't mutate services here so we avoid calling
+    # async_services which makes a copy of every services
+    # dict.
+    services = hass.services._services  # pylint: disable=protected-access
```

Comment:
I understand that we're trying to improve performance but circumventing the design set for an object will make it harder to reason about the use of that object.

If we want to have a faster way of accessing a feature of an object we should make a proper interface for that.

Reply:
I can add a new method but I was a bit worried someone would misuse it and mutate the service data but I think I might be overly concerned there.  I'm not sure why the original design copied every service as I can't seems to find where that is needed.

Reply:
I've analyzed this a bit and it seems like there `async_services` is almost always used for `hass.services.async_services()[DOMAIN]` 

The reason we make a copy is we iterate the services to remove them.

I'll propose a PR shortly to adapt the services api to better fit the actual use cases.

Reply:
https://github.com/home-assistant/core/pull/108671